### PR TITLE
Make CI error if a function has no documentation

### PR DESCRIPTION
**Pull request by alamb**
Draft until we have completely migrated all function documentation. The CI will fail until all currently existing functions are documented

## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/12872

## Rationale for this change

The idea here is to ensure that all functions are documented, and ensure this is the case via a CI test.

## What changes are included in this PR?

Changes:
make the ./dev/update_function_docs.sh script error if there are undocumented functions

Here is an example failure:

```
(venv) andrewlamb@Andrews-MacBook-Pro-2:~/Software/datafusion$ ./dev/update_function_docs.sh
/Users/andrewlamb/Software/datafusion
Inserting header
Running CLI and inserting aggregate function docs table
   Compiling datafusion v42.0.0 (/Users/andrewlamb/Software/datafusion/datafusion/core)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.77s
     Running `target/debug/print_functions_docs aggregate`
INFO: The following functions do not have documentation:
  - regr_count
  - regr_avgx
  - regr_syy
  - regr_intercept
  - regr_slope
  - regr_sxy
  - regr_r2
  - regr_sxx
  - regr_avgy
Error: NotImplemented("Some functions do not have documentation. Please implement `documentation` for: {\"regr_count\", \"regr_avgx\", \"regr_syy\", \"regr_intercept\", \"regr_slope\", \"regr_sxy\", \"regr_r2\", \"regr_sxx\", \"regr_avgy\"}")
(venv) andrewlamb@Andrews-MacBook-Pro-2:~/Software/datafusion$
```
## Are these changes tested?
Yes, I tested it manually and here is an example of it failing

### CODE REVIEW

**Comment by alamb** [comment]
Example CI failure: https://github.com/apache/datafusion/actions/runs/11344725082/job/31550591277?pr=12938

File: datafusion/core/src/bin/print_functions_docs.rs [hunk 1]
```diff
@@ -48,12 +49,13 @@ fn main() {
         _ => {
             panic!("Unknown function type: {}", function_type)
         }
-    };
+    }?;
 
     println!("{docs}");
```
Comment (comphead) [inline datafusion/core/src/bin/print_functions_docs.rs:54]: do we need this println?

**Review by comphead — APPROVED** [review]
lgtm thanks @alamb

**Review by timsaucer — APPROVED** [review]
Nice addition!

File: datafusion/core/src/bin/print_functions_docs.rs (hunk 1 above)
Comment (Omega359) [reply datafusion/core/src/bin/print_functions_docs.rs:54]: Yes. This is how the shell script that builds the .md files retrieves the text. see https://github.com/apache/datafusion/blob/main/dev/update_function_docs.sh for impl.

**Comment by alamb** [comment]
Bam ! Thank you for the reviews @comphead @Omega359 and @timsaucer  -- We are so close to getting the docs all migrated over

**Pull request merged**